### PR TITLE
fix: do not re-sort contracting axes for generalized dot product

### LINKED ISSUE #612: Incorrect dot result with BinaryBackend
```elixir
a = Nx.tensor([[0.0, 1.0, 2.0], [3.0, 4.0, 5.0]])
b = Nx.tensor([[0.0, 1.0], [2.0, 3.0], [4.0, 5.0]])
c = Nx.dot(a, [0, 1], b, [1, 0])
IO.inspect(c)
```
```
#Nx.Tensor<
  f32
  55.0
>
```

With pytorch(and torchx backend):

```python
import torch

a = torch.tensor([[0.0, 1.0, 2.0], [3.0, 4.0, 5.0]])
b = torch.tensor([[0.0, 1.0], [2.0, 3.0], [4.0, 5.0]])
c = torch.tensordot(a, b, dims=([0, 1],[1, 0]))
print(c)
```

```
tensor(50.)
```

Looks like BinaryBackend ignores tensor shapes and computes as if they are vectors: `0*0 + 1*1 + 2*2 + 3*3 + 4*4 + 5*5 = 55`

Also because of this [at least one test](https://github.com/elixir-nx/nx/blob/efed8861da20a256e6d328f61cd7c01bf4090620/nx/test/nx_test.exs#L1118) is incorrect

### PR DESCRIPTION
**Pull request by polvalente**
closes #612

### CODE REVIEW

File: nx/lib/nx/binary_backend.ex [hunk 1]
```diff
@@ -521,12 +521,36 @@ defmodule Nx.BinaryBackend do
...
     end)
   end
 
+  defp bin_dot_transpose_contract_axes(tensor, contract_axes) do
+    remaining_axes =
+      contract_axes
+      |> Enum.sort(:desc)
+      |> Enum.reduce(Nx.axes(tensor), &List.delete_at(&2, &1))
+
+    transposed = Nx.transpose(tensor, axes: contract_axes ++ remaining_axes)
```
Comment (polvalente) [inline nx/lib/nx/binary_backend.ex:542]: @josevalim is it bad that I'm calling Nx.transpose and Nx.reshape here?
Should I call the BinaryBackend equivalents?

Comment (josevalim) [reply nx/lib/nx/binary_backend.ex:546]: That would be preferred, yes!

File: nx/lib/nx/binary_backend.ex [hunk 2]
```diff
@@ -521,12 +521,40 @@ defmodule Nx.BinaryBackend do
...
-    bin_zip_reduce(left, contract_axes1, right, contract_axes2, type, 0, fn lhs, rhs, acc ->
+    {left, left_contract_axes} = bin_dot_transpose_contract_axes(left, contract_axes1)
+
+    {right, right_contract_axes} = bin_dot_transpose_contract_axes(right, contract_axes2)
+
+    bin_zip_reduce(left, left_contract_axes, right, right_contract_axes, type, 0, fn lhs,
+                                                                                     rhs,
+                                                                                     acc ->
       res = binary_to_number(lhs, t1) * binary_to_number(rhs, t2) + acc
       {res, res}
```
Comment (josevalim) [inline nx/lib/nx/binary_backend.ex:532]:
```suggestion
    bin_zip_reduce(left, left_contract_axes, right, right_contract_axes, type, 0,
      fn lhs, rhs, acc ->
        res = binary_to_number(lhs, t1) * binary_to_number(rhs, t2) + acc
        {res, res}
```

File: nx/lib/nx.ex [hunk 3]
```diff
@@ -7172,6 +7172,15 @@ defmodule Nx do
...
 
+      iex> t1 = Nx.tensor([[0.0, 1.0, 2.0], [3.0, 4.0, 5.0]])
+      iex> t2 = Nx.tensor([[0.0, 1.0], [2.0, 3.0], [4.0, 5.0]])
+      iex> Nx.dot(t1, [0, 1], t2, [1, 0])
+      #Nx.Tensor<
+        f32
+        50.0
+      >
+
+
```
Comment (josevalim) [inline nx/lib/nx.ex:7183]:
```suggestion
```

File: nx/lib/nx/binary_backend.ex [hunk 4]
```diff
@@ -521,12 +521,51 @@ defmodule Nx.BinaryBackend do
...
     end)
   end
 
+  defp bin_dot_transpose_contract_axes(tensor, contract_axes) do
+    # The intution here is that we can pre-condense the contracting axes into a
+    # single dimension, which will then be contracted through bin_zip_reduce below.
+    # This takes a shape {a, m, n, b} which contracts on m, n and turns it into
+    # {m * n, a, b}, contracting on the first dimension.
+
+    axes = Enum.to_list(0..(tuple_size(tensor.shape) - 1)//1)
```
Comment (josevalim) [inline nx/lib/nx/binary_backend.ex:541]:
```suggestion
    axes = Nx.axes(tensor)
```

File: nx/lib/nx/binary_backend.ex [hunk 5]
```diff
@@ -521,12 +521,51 @@ defmodule Nx.BinaryBackend do
...
+
+    {contracted, kept} =
+      transposed.shape
+      |> Tuple.to_list()
+      |> Enum.split(length(contract_axes))
+
+    reduced_shape = List.to_tuple([Enum.product(contracted) | kept])
+
+    {%{transposed | shape: reduced_shape, names: List.duplicate(nil, tuple_size(reduced_shape))},
+     [0]}
```
Comment (josevalim) [inline nx/lib/nx/binary_backend.ex:566]: I actually think the most efficient implementation would be to move them to lower axis, because we can then traverse them with fewer jumps. Sorry, I should have noticed this sooner.

Comment (polvalente) [reply nx/lib/nx/binary_backend.ex:566]: You mean the "-1" axis?

Comment (josevalim) [reply nx/lib/nx/binary_backend.ex:566]: Yes, the -1.

**Review by josevalim — APPROVED** [review]

File: nx/lib/nx/binary_backend.ex [hunk 6]
```diff
@@ -521,12 +521,55 @@ defmodule Nx.BinaryBackend do
...
+        res = binary_to_number(lhs, t1) * binary_to_number(rhs, t2) + acc
+        {res, res}
     end)
   end
 
+  defp bin_dot_transpose_contract_axes(tensor, contract_axes) do
+    # The intution here is that we can pre-condense the contracting axes into a
+    # single dimension, which will then be contracted through bin_zip_reduce below.
+    # This takes a shape {a, m, n, b} which contracts on m, n and turns it into
+    # {a, b, m * n}, contracting on the last dimension.
```
Comment (josevalim) [inline nx/lib/nx/binary_backend.ex:539]:
```suggestion
    # {a, b, m * n}, contracting on the last dimension. This is necessary because bin_zip_reduce
    # and aggregate_axes are order independent but dot depends on the axes order.
```
